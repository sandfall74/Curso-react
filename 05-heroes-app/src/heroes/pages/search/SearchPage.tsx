import { CustomJumbotron } from '@/components/custom/CustomJumbotron';
import { HeroStats } from '@/heroes/components/HeroStats';
import { SearchControls } from './ui/SearchControls';
import { CustomBreadcrumbs } from '@/components/custom/CustomBreadcrumbs';
import { HeroGrid } from '@/heroes/components/HeroGrid';
import { useQuery } from '@tanstack/react-query';
import { useSearchParams } from 'react-router';
import { searchHeroesAction } from '@/heroes/actions/search-heros.action';

export const SearchPage = () => {
  const [searchParams] = useSearchParams();

  const name = searchParams.get('name') ?? undefined;
  const strength = searchParams.get('strength') ?? undefined;

  const { data: heroes = [] } = useQuery({
    queryKey: ['search', { name, strength }],
    queryFn: () => searchHeroesAction({ name, strength }),
    staleTime: 1000 * 60 * 5, // 5 minutos
  });

  return (
    <>
      <CustomJumbotron
        title="Búsqueda de SuperHéroes"
        description="Descubre, explora y administra super héroes y villanos"
      />

      <CustomBreadcrumbs
        currentPage="Buscador de héroes"
        // breadcrumbs={[
        //   { label: 'Home1', to: '/' },
        //   { label: 'Home2', to: '/' },
        //   { label: 'Home3', to: '/' },
        // ]}
      />

      {/* Stats Dashboard */}
      <HeroStats />

      {/* Filter and search */}
      <SearchControls />

      {/*  */}

      <HeroGrid heroes={heroes} />
    </>
  );
};

export default SearchPage;
