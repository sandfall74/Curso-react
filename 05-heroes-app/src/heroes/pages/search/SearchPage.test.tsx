import { render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, test, vi } from 'vitest';
import SearchPage from './SearchPage';
import { MemoryRouter } from 'react-router';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { searchHeroesAction } from '@/heroes/actions/search-heros.action';
import type { Hero } from '@/heroes/types/hero.interface';

vi.mock('@/heroes/actions/search-heros.action');
const mockSearchHeroesAction = vi.mocked(searchHeroesAction);

vi.mock('@/components/custom/CustomJumbotron', () => ({
  CustomJumbotron: () => <div data-testid="custom-jumbotron"></div>,
}));

vi.mock('./ui/SearchControls', () => ({
  SearchControls: () => <div data-testid="search-controls"></div>,
}));

vi.mock('@/heroes/components/HeroGrid', () => ({
  HeroGrid: ({ heroes }: { heroes: Hero[] }) => (
    <div data-testid="hero-grid">
      {heroes.map((hero) => (
        <div key={hero.id}>{hero.name}</div>
      ))}
    </div>
  ),
}));

const queryClient = new QueryClient();

const renderSearchPage = (initialEntries: string[] = ['/']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <QueryClientProvider client={queryClient}>
        <SearchPage />
      </QueryClientProvider>
    </MemoryRouter>
  );
};

describe('SearchPage', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('should render SearchPage with default values', () => {
    const { container } = renderSearchPage();

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: undefined,
    });

    expect(container).toMatchSnapshot();
  });

  test('should call search action with name parameter', () => {
    const { container } = renderSearchPage(['/search?name=superman']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'superman',
      strength: undefined,
    });

    expect(container).toMatchSnapshot();
  });

  test('should call search action with strength parameter', () => {
    const { container } = renderSearchPage(['/search?strength=6']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: '6',
    });

    expect(container).toMatchSnapshot();
  });

  test('should call search action with strength and name parameters', () => {
    const { container } = renderSearchPage(['/search?strength=8&name=batman']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'batman',
      strength: '8',
    });

    expect(container).toMatchSnapshot();
  });

  test('should render HeroGrid with search results', async () => {
    const mockHeroes = [
      { id: '1', name: 'Clark Kent' } as unknown as Hero,
      { id: '2', name: 'Bruce Wayne' } as unknown as Hero,
    ];

    mockSearchHeroesAction.mockResolvedValue(mockHeroes);

    renderSearchPage();

    await waitFor(() => {
      expect(screen.getByText('Clark Kent')).toBeDefined();
      expect(screen.getByText('Bruce Wayne')).toBeDefined();
    });
  });
});
