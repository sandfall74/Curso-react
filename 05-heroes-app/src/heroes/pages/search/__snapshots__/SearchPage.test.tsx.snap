// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > should call search action with name parameter 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
        >
          Buscador de héroes
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with strength and name parameters 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
        >
          Buscador de héroes
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with strength parameter 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
        >
          Buscador de héroes
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should render SearchPage with default values 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
        >
          Buscador de héroes
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;
